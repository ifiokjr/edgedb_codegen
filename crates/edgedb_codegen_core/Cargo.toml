[package]
name = "edgedb_codegen_core"
version = { workspace = true }
authors = { workspace = true }
categories = ["database"]
edition = { workspace = true }
homepage = { workspace = true }
keywords = ["edgedb", "database", "typesafe", "checked", "macros"]
license = { workspace = true }
readme = "readme.md"
repository = { workspace = true }
rust-version = { workspace = true }
description = "Utilities for codegen in the `edgedb_codegen` crate."

[dependencies]
check_keyword = { workspace = true }
edgedb-errors = { workspace = true }
edgedb-protocol = { workspace = true, features = ["all-types", "with-serde"] }
edgedb-tokio = { workspace = true, features = ["unstable", "derive"] }
heck = { workspace = true }
log = { workspace = true }
prettyplease = { workspace = true }
proc-macro2 = { workspace = true }
quote = { workspace = true }
syn = { workspace = true }
thiserror = { workspace = true }
tokio = { workspace = true, features = ["macros", "rt-multi-thread", "process"] }
typed-builder = { workspace = true }

[dev-dependencies]
assert2 = { workspace = true }
tokio = { workspace = true, features = ["time", "test-util"] }

[features]
with_bigint = []
with_bigdecimal = []
with_chrono = []
with_all = ["with_bigint", "with_bigdecimal", "with_chrono"]

[lints]
workspace = true
