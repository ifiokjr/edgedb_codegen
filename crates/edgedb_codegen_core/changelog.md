# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.2](https://github.com/ifiokjr/edgedb_codegen/compare/edgedb_codegen_core@v0.1.1...edgedb_codegen_core@v0.1.2) - 2024-08-27

### <!-- 0 -->🎉 Added
- *(edgedb_codegen_core)* add `prettyplease` format option

### <!-- 1 -->🐛 Bug Fixes
- increase minimum rust version

### <!-- 7 -->⚙️ Miscellaneous Tasks
- update coverage files
- support testing multiple rust versions

## [0.1.1](https://github.com/ifiokjr/edgedb_codegen/compare/edgedb_codegen_core@0.1.0...edgedb_codegen_core@0.1.1) - 2024-08-26

### <!-- 0 -->🎉 Added
- add support for `Range` types

### <!-- 7 -->⚙️ Miscellaneous Tasks
- improve changelog generation

## [0.1.0](https://github.com/ifiokjr/edgedb_codegen/releases/tag/edgedb_codegen_core-v0.1.0) - 2024-08-25

### 🎉 Added

- Initial release
