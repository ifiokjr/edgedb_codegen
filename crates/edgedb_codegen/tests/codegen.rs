use std::path::PathBuf;
use std::thread;

use edgedb_codegen_core::*;
use proc_macro2::Span;
use rstest::fixture;
use rstest::rstest;

macro_rules! set_snapshot_suffix {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_suffix(format!($($expr,)*));
        let _guard = settings.bind_to_scope();
    }
}

#[fixture]
pub fn testname() -> String {
	thread::current()
		.name()
		.unwrap()
		.split("::")
		.last()
		.unwrap()
		.to_string()
}

#[rstest]
#[case::str("select 'i ❤️ edgedb'")]
#[case::bool("select true")]
#[case::int64("select 3")]
#[case::float64("select 314e-2")]
#[case::bigint("select 42n")]
#[case::decimal("select 42e+100n")]
#[case::uuid("select <uuid>'a5ea6360-75bd-4c20-b69c-8f317b0d2857'")]
#[case::datetime("select <datetime>'1999-03-31T15:17:00Z'")]
#[case::duration("select <duration>'45.6 seconds'")]
#[case::array("select [1, 2, 3]")]
#[case::tuple("select ('Apple', 7, true)")]
#[case::named_tuple("select (fruit := 'Apple', quantity := 3.14, fresh := true)")]
#[case::set(r#"select {"set", "of", "strings"}"#)]
#[case::empty_set("select <int64>{}")]
#[case::input_shape(
	"select { my_string := RelationshipType.Follow, my_number := 42, several_numbers := {1, 2, \
	 3}, array := [1, 2, 3] };"
)]
#[case::input_shape_with_args(r#"select { hello := "world", custom := <str>$custom }"#)]
#[case::object_shape("select Team {**}")]
#[case::object_shape_with_args(
	"select Team {**} filter .name like <str>$starts_with ++ '%' and .description like '%' ++ \
	 <str>$ends_with;"
)]
#[case::types_query(TYPES_QUERY)]
// #[case::range("select range(0, 10)")] // TODO: `Range` doesn't implement `Queryable` yet.
// #[case::bytes("select b'bina\\x01ry'")] // TODO: bytes don't implement `DecodeScalar` yet.
#[tokio::test]
#[rustversion::attr(not(nightly), ignore = "requires nightly")]
async fn codegen_literals(testname: String, #[case] query: &str) -> Result<()> {
	set_snapshot_suffix!("{}", testname);

	let relative_path = format!("tests/compile/codegen/{testname}.rs");
	let descriptor = get_descriptor(query).await?;
	let code = generate_rust_from_query(&descriptor, "example", query)?;
	let content = rustfmt(&code.to_string()).await?;

	// Check that the snapshot hasn't changed.
	insta::assert_snapshot!(&content);

	// Ensure that the produced rust is valid.
	prepare_compile_test(&content, &relative_path).await?;

	Ok(())
}

#[rstest]
#[case("insert_user")]
#[case("remove_user")]
#[tokio::test]
#[rustversion::attr(not(nightly), ignore = "requires nightly")]
async fn codegen_files(#[case] path: &str) -> Result<()> {
	set_snapshot_suffix!("{}", path);

	let query_path = resolve_path(format!("queries/{path}.edgeql"), Span::call_site())?;
	let relative_path = format!("tests/compile/codegen/{path}.rs");
	let query = tokio::fs::read_to_string(&query_path).await?;
	let descriptor = get_descriptor(&query).await?;
	let code = generate_rust_from_query(&descriptor, "example", &query)?;
	let content = rustfmt(&code.to_string()).await?;

	// Check that the snapshot hasn't changed.
	insta::assert_snapshot!(&content);

	// Ensure that the produced rust is valid.
	prepare_compile_test(&content, &relative_path).await?;

	Ok(())
}

const CRATE_DIR: &str = env!("CARGO_MANIFEST_DIR");

async fn prepare_compile_test(content: &str, relative_path: &str) -> Result<()> {
	let is_ci = std::env::var("CI")
		.ok()
		.is_some_and(|v| ["1", "true"].contains(&v.as_str()));

	let path = PathBuf::from(CRATE_DIR).join(relative_path);
	let generated = generate_contents(content).await?;

	let should_update = match tokio::fs::read_to_string(&path).await {
		Ok(current) => current != generated,
		Err(_) => true,
	};

	if should_update {
		assert2::assert!(!is_ci, "attempted updating compilation tests in CI");
		tokio::fs::write(&path, generated).await?;
	}

	Ok(())
}

async fn generate_contents(content: &str) -> Result<String> {
	let updated = format!("fn main() {{\n{content}\n}}");
	Ok(rustfmt(&updated).await?)
}
