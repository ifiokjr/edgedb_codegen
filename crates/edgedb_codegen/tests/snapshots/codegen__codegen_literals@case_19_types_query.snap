---
source: crates/edgedb_codegen/tests/codegen.rs
expression: "&content"
---
pub mod example {
    use ::edgedb_codegen::exports as e;
    /// Execute the desired query.
    #[cfg(feature = "query")]
    pub async fn query(
        client: &e::edgedb_tokio::Client,
    ) -> core::result::Result<Vec<Output>, e::edgedb_errors::Error> {
        client.query(QUERY, &()).await
    }
    /// Compose the query as part of a larger transaction.
    #[cfg(feature = "query")]
    pub async fn transaction(
        conn: &mut e::edgedb_tokio::Transaction,
    ) -> core::result::Result<Vec<Output>, e::edgedb_errors::Error> {
        conn.query(QUERY, &()).await
    }
    pub type Input = ();
    #[derive(Clone, Debug)]
    #[cfg_attr(feature = "query", derive(e::edgedb_derive::Queryable))]
    #[cfg_attr(feature = "serde", derive(e::serde::Serialize, e::serde::Deserialize))]
    pub struct OutputBasesSet {
        pub id: e::uuid::Uuid,
    }
    #[derive(Clone, Debug)]
    #[cfg_attr(feature = "query", derive(e::edgedb_derive::Queryable))]
    #[cfg_attr(feature = "serde", derive(e::serde::Serialize, e::serde::Deserialize))]
    pub struct OutputUnionOfSet {
        pub id: e::uuid::Uuid,
    }
    #[derive(Clone, Debug)]
    #[cfg_attr(feature = "query", derive(e::edgedb_derive::Queryable))]
    #[cfg_attr(feature = "serde", derive(e::serde::Serialize, e::serde::Deserialize))]
    pub struct OutputIntersectionOfSet {
        pub id: e::uuid::Uuid,
    }
    #[derive(Clone, Debug)]
    #[cfg_attr(feature = "query", derive(e::edgedb_derive::Queryable))]
    #[cfg_attr(feature = "serde", derive(e::serde::Serialize, e::serde::Deserialize))]
    pub struct OutputPointersSetPointersSet {
        pub card: Option<String>,
        pub name: String,
        pub target_id: Option<e::uuid::Uuid>,
        pub kind: String,
        pub is_computed: Option<bool>,
        pub is_readonly: Option<bool>,
    }
    #[derive(Clone, Debug)]
    #[cfg_attr(feature = "query", derive(e::edgedb_derive::Queryable))]
    #[cfg_attr(feature = "serde", derive(e::serde::Serialize, e::serde::Deserialize))]
    pub struct OutputPointersSet {
        pub card: Option<String>,
        pub name: String,
        pub target_id: Option<e::uuid::Uuid>,
        pub kind: String,
        pub is_exclusive: bool,
        pub is_computed: Option<bool>,
        pub is_readonly: Option<bool>,
        pub has_default: bool,
        pub pointers: Vec<OutputPointersSetPointersSet>,
    }
    #[derive(Clone, Debug)]
    #[cfg_attr(feature = "query", derive(e::edgedb_derive::Queryable))]
    #[cfg_attr(feature = "serde", derive(e::serde::Serialize, e::serde::Deserialize))]
    pub struct OutputExclusivesSet {
        pub target: Option<String>,
    }
    #[derive(Clone, Debug)]
    #[cfg_attr(feature = "query", derive(e::edgedb_derive::Queryable))]
    #[cfg_attr(feature = "serde", derive(e::serde::Serialize, e::serde::Deserialize))]
    pub struct OutputBacklinksSet {
        pub card: String,
        pub name: String,
        pub stub: String,
        pub target_id: Option<e::uuid::Uuid>,
        pub kind: String,
        pub is_exclusive: Option<bool>,
    }
    #[derive(Clone, Debug)]
    #[cfg_attr(feature = "query", derive(e::edgedb_derive::Queryable))]
    #[cfg_attr(feature = "serde", derive(e::serde::Serialize, e::serde::Deserialize))]
    pub struct OutputBacklinkStubsArray {
        pub card: String,
        pub name: String,
        pub target_id: Option<e::uuid::Uuid>,
        pub kind: String,
        pub is_exclusive: bool,
    }
    #[derive(Clone, Debug)]
    #[cfg_attr(feature = "query", derive(e::edgedb_derive::Queryable))]
    #[cfg_attr(feature = "serde", derive(e::serde::Serialize, e::serde::Deserialize))]
    pub struct OutputTupleElementsSet {
        pub target_id: e::uuid::Uuid,
        pub name: Option<String>,
    }
    #[derive(Clone, Debug)]
    #[cfg_attr(feature = "query", derive(e::edgedb_derive::Queryable))]
    #[cfg_attr(feature = "serde", derive(e::serde::Serialize, e::serde::Deserialize))]
    pub struct Output {
        pub id: e::uuid::Uuid,
        pub name: String,
        pub is_abstract: Option<bool>,
        pub kind: String,
        pub enum_values: Option<Vec<String>>,
        pub is_seq: bool,
        pub material_id: Option<e::uuid::Uuid>,
        pub bases: Vec<OutputBasesSet>,
        pub union_of: Vec<OutputUnionOfSet>,
        pub intersection_of: Vec<OutputIntersectionOfSet>,
        pub pointers: Vec<OutputPointersSet>,
        pub exclusives: Vec<OutputExclusivesSet>,
        pub backlinks: Vec<OutputBacklinksSet>,
        pub backlink_stubs: Vec<OutputBacklinkStubsArray>,
        pub array_element_id: Option<e::uuid::Uuid>,
        pub tuple_elements: Vec<OutputTupleElementsSet>,
        pub multirange_element_id: Option<e::uuid::Uuid>,
    }
    /// The original query string provided to the macro. Can be reused in your codebase.
    pub const QUERY: &str = "WITH\n  MODULE schema,\n  material_scalars := (\n    SELECT ScalarType\n    FILTER NOT .abstract\n       AND NOT EXISTS .enum_values\n       AND NOT EXISTS (SELECT .ancestors FILTER NOT .abstract)\n  )\n\n\tSELECT Type {\n\t  id,\n\t  name :=\n\t    array_join(array_agg([IS ObjectType].union_of.name), ' | ')\n\t    IF EXISTS [IS ObjectType].union_of\n\t    ELSE .name,\n\t  is_abstract := .abstract,\n\n\t  kind := 'object' IF Type IS ObjectType ELSE\n\t          'scalar' IF Type IS ScalarType ELSE\n\t          'array' IF Type IS Array ELSE\n\t          'tuple' IF Type IS Tuple ELSE\n\t          'multirange' IF Type IS MultiRange ELSE\n\t          'unknown',\n\n\t  [IS ScalarType].enum_values,\n\t  is_seq := 'std::sequence' in [IS ScalarType].ancestors.name,\n\t  # for sequence (abstract type that has non-abstract ancestor)\n\t  single material_id := (\n\t    SELECT x := Type[IS ScalarType].ancestors\n\t    FILTER x IN material_scalars\n\t    LIMIT 1\n\t  ).id,\n\n\t  [IS InheritingObject].bases: {\n\t    id\n\t  } ORDER BY @index ASC,\n\n\t  [IS ObjectType].union_of,\n\t  [IS ObjectType].intersection_of,\n\t  [IS ObjectType].pointers: {\n\t    card := ('One' IF .required ELSE 'AtMostOne') IF <str>.cardinality = 'One' ELSE ('AtLeastOne' IF .required ELSE 'Many'),\n\t    name,\n\t    target_id := .target.id,\n\t    kind := 'link' IF .__type__.name = 'schema::Link' ELSE 'property',\n\t    is_exclusive := exists (select .constraints filter .name = 'std::exclusive'),\n\t    is_computed := len(.computed_fields) != 0,\n\t    is_readonly := .readonly,\n\t    has_default := EXISTS .default or ('std::sequence' in .target[IS ScalarType].ancestors.name),\n\t    [IS Link].pointers: {\n\t      card := ('One' IF .required ELSE 'AtMostOne') IF <str>.cardinality = \"One\" ELSE ('AtLeastOne' IF .required ELSE 'Many'),\n\t      name := '@' ++ .name,\n\t      target_id := .target.id,\n\t      kind := 'link' IF .__type__.name = 'schema::Link' ELSE 'property',\n\t      is_computed := len(.computed_fields) != 0,\n\t      is_readonly := .readonly\n\t    } filter .name != '@source' and .name != '@target',\n\t  } FILTER @is_owned,\n\t  exclusives := assert_distinct((\n\t    [is schema::ObjectType].constraints\n\t    union\n\t    [is schema::ObjectType].pointers.constraints\n\t  ) {\n\t    target := (.subject[is schema::Property].name ?? .subject[is schema::Link].name ?? .subjectexpr)\n\t  } filter .name = 'std::exclusive'),\n\t  backlinks := (\n\t     SELECT DETACHED Link\n\t     FILTER .target = Type\n\t       AND NOT EXISTS .source[IS ObjectType].union_of\n\t    ) {\n\t    card := 'AtMostOne'\n\t      IF\n\t      EXISTS (select .constraints filter .name = 'std::exclusive')\n\t      ELSE\n\t      'Many',\n\t    name := '<' ++ .name ++ '[is ' ++ assert_exists(.source.name) ++ ']',\n\t    stub := .name,\n\t    target_id := .source.id,\n\t    kind := 'link',\n\t    is_exclusive := (EXISTS (select .constraints filter .name = 'std::exclusive')) AND <str>.cardinality = 'One',\n\t  },\n\t  backlink_stubs := array_agg((\n\t    WITH\n\t      stubs := DISTINCT (SELECT DETACHED Link FILTER .target = Type).name,\n\t      baseObjectId := (SELECT DETACHED ObjectType FILTER .name = 'std::BaseObject' LIMIT 1).id\n\t    FOR stub in { stubs }\n\t    UNION (\n\t      SELECT {\n\t        card := 'Many',\n\t        name := '<' ++ stub,\n\t        target_id := baseObjectId,\n\t        kind := 'link',\n\t        is_exclusive := false,\n\t      }\n\t    )\n\t  )),\n\t  array_element_id := [IS Array].element_type.id,\n\n\t  tuple_elements := (SELECT [IS Tuple].element_types {\n\t    target_id := .type.id,\n\t    name\n\t  } ORDER BY @index ASC),\n\t\t multirange_element_id := [IS MultiRange].element_type.id,\n\t}\nORDER BY .name;\n";
}
